src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -285,36 +325,94 @@ private void ProcessTraceroute(string targetNameOrAddress)
         {
             byte[] buffer = GetSendBuffer(BufferSize);
 
-            string resolvedTargetName;
-            IPAddress targetAddress;
-            if (!InitProcessPing(targetNameOrAddress, out resolvedTargetName, out targetAddress))
+            if (!InitProcessPing(targetNameOrAddress, out string resolvedTargetName, out IPAddress targetAddress))
             {
                 return;
             }
 
-            TraceRouteResult traceRouteResult = new TraceRouteResult(Source, targetAddress, resolvedTargetName);
-
             int currentHop = 1;
             PingOptions pingOptions = new PingOptions(currentHop, DontFragment.IsPresent);
-            PingReply reply = null;
+            PingReply reply;
+            PingReply discoveryReply;
             int timeout = TimeoutSeconds * 1000;
+            Stopwatch timer = new Stopwatch();
 
+            IPAddress hopAddress;
             do
             {
-                TraceRouteReply traceRouteReply = new TraceRouteReply();
-
-                pingOptions.Ttl = traceRouteReply.Hop = currentHop;
-                currentHop++;
-
-                // In the specific case we don't use 'Count' property.
+                // Clear the stored router name for every hop
+                string routerName = null;
+                pingOptions.Ttl = currentHop;
+
+#if !UNIX
+                // Get intermediate hop target. This needs to be done first, so that we can target it properly
+                // and get useful responses.
+                var discoveryAttempts = 0;
+                bool addressIsValid = false;
+                do
+                {
+                    discoveryReply = SendCancellablePing(targetAddress, timeout, buffer, pingOptions);
+                    discoveryAttempts++;
+                    addressIsValid = !(discoveryReply.Address.Equals(IPAddress.Any)
+                        || discoveryReply.Address.Equals(IPAddress.IPv6Any));
+                }
+                while (discoveryAttempts <= 3 && addressIsValid);
+
+                // If we aren't able to get a valid address, just re-target the final destination of the trace.
+                hopAddress = addressIsValid ? discoveryReply.Address : targetAddress;
+#endif
```

Comment:
Why not just `#else`?

Reply:
Because I don't use compiler directives often and didn't know `#else` was allowed! Will fix. 🙂